composer/trainer/checkpoint.py

```diff
@@ -135,25 +157,78 @@ def load_checkpoint(self, state: State):
+                # and only rank zero has this file
+                extracted_checkpoint_folder = None
+                mosaic_checkpoint_filepath = rank_zero_checkpoint_archive_filepath
             else:
-                extracted_checkpoint_folder = os.path.join(checkpoint_folder, "checkpoint")
-                with tarfile.open(checkpoint_archive_filepath) as tarball:
-                    tarball.extractall(extracted_checkpoint_folder)
+                extracted_checkpoint_folder = os.path.join(node_checkpoint_folder, "checkpoint")
                 mosaic_checkpoint_filepath = os.path.join(extracted_checkpoint_folder, _MOSAIC_STATES_FILENAME)
 
+            if dist.get_local_rank() == 0:
+                # every NODE needs the GLOBAL rank zero checkpoint
+                self._retrieve_checkpoint(destination_filepath=rank_zero_checkpoint_archive_filepath,
+                                          rank=dist.get_global_rank(),
+                                          ignore_not_found_errors=False)
+
+                if extracted_checkpoint_folder is not None:
+                    try:
+                        with tarfile.open(rank_zero_checkpoint_archive_filepath) as tarball:
+                            tarball.extractall(extracted_checkpoint_folder)
+                    except FileNotFoundError as e:
+                        checkpoint_name = self.hparams.checkpoint.format(rank=dist.get_global_rank())
+                        raise RuntimeError(f"Unable to retrieve checkpoint {checkpoint_name}") from e
+
+            if rank_zero_checkpoint_archive_filepath != rank_n_checkpoint_archive_filepath:
+                # every RANK needs ITS OWN checkpoint.
+                # But, the  global rank zero is a special case -- these files are the same!
+                assert dist.get_global_rank() != 0, "invariant violation"
+
+                # Allowing not-found errors to be ignored as sometimes there won't be rank-local checkpoints
+                # (e.g. when not using deepspeed)
+                self._retrieve_checkpoint(destination_filepath=rank_n_checkpoint_archive_filepath,
+                                          rank=dist.get_global_rank(),
+                                          ignore_not_found_errors=True)
+
+                if extracted_checkpoint_folder is not None:
+                    # it's an archive and needs to be extracted
+                    with tarfile.open(rank_n_checkpoint_archive_filepath) as tarball:
+                        tarball.extractall(extracted_checkpoint_folder)
+
+            # Wait for all checkpoints on the node to finish downloading
+            # technically, for this barrier, we only need to wait for the
+            # global rank zero checkpoint, but putting the barrier here
+            # allows for other ranks to begin downloading their checkpoints
+            dist.barrier()
+
+            # Now, all ranks load the checkpoint that local rank zero downloaded
```

Comment:
This function is big enough that it's very difficult to read and keep track of indentation. Could you break it up into smaller subfunctions? It seems like downloading/unpacking the checkpoints and loading the checkpoints could belong to separate functions.

Reply:
Good point; refactored the function for readability.